<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
  <title>SEMPHY (VERSION 2.00)</title>
</head>
<body dir="ltr">
<p> <span style="font-family: Arial;"><span
 style="text-decoration: underline;"><span style="font-weight: bold;">SEMPHY
(Version 2.00)<br>
</span></span></span>This manual covers the following topics:<span
 style="font-family: Arial;"><span style="text-decoration: underline;"><span
 style="font-weight: bold;"><br>
</span></span></span><span dir="ltr"><span style="font-style: italic;"></span></span>
</p>
<ul>
  <li><span style="font-weight: bold;">Neighbor
Joining (NJ) trees</span></li>
  <li><span style="font-weight: bold;">Maximum Likelihood (ML) trees
using SEMPHY</span></li>
  <li><span style="font-weight: bold;">List of options and parameters</span></li>
  <li><span style="font-weight: bold;">Running on large datasets</span></li>
</ul>
<p><span style="font-family: Times New Roman,Times,serif;">All uses of
SEMPHY are by calling the executable semphy from a command prompt with
some parameters:</span><br>
&nbsp;&nbsp;
<span style="font-family: Courier New,Courier,monospace;">semphy
[parameters...]</span></p>
<p>Here we give a few examples of using SEMPHY for the most common
tasks.&nbsp; Below is a table with the full list of options, which is
also available by typing '<span
 style="font-family: Courier New,Courier,monospace;">sepmhy -h</span>'
at the command prompt.</p>
<p><span style="font-weight: bold; text-decoration: underline;">Neighbor
Joining (NJ) trees</span><span style="font-style: italic;"><br>
</span>Running standard NJ on<span
 style="font-weight: bold; font-style: italic; color: rgb(153, 51, 153);">
protein
sequences</span> using the JTT replacement matrix:<br>
&nbsp;&nbsp; <span style="font-family: Courier New,Courier,monospace;">semphy
-s prots.fasta -o out.txt -T prots.tree -l log.txt -a 20 --jtt -J -H</span><br>
(Meaning:&nbsp; Input is a fasta sequence file and ouput is written to
three files (general output,
tree file and log file).&nbsp; Use alphabet of 20, i.e. amino
acids.&nbsp;
Use the JTT matrix.&nbsp; Do Neighbor Joining, with a homogeneous rates
model)<br>
</p>
<p>Same thing, but using the <span
 style="font-style: italic; font-weight: bold; color: rgb(153, 51, 153);">new
iterative NJ</span> method with Bayesian estimation of the rate at each
site (i.e. "posterior" estimates of the rates. -O requests optimization
of rate parameters)<br>
&nbsp;&nbsp; <span style="font-family: Courier New,Courier,monospace;">semphy
-s prots.fasta -o out.txt -T prots.tree </span><span
 style="font-family: Courier New,Courier,monospace;">-l log.txt </span><span
 style="font-family: Courier New,Courier,monospace;">-a 20 --jtt
--posterior</span><span
 style="font-family: Courier New,Courier,monospace;">DTME -O<br>
</span></p>
<p>Running iterative NJ on <span
 style="font-style: italic; font-weight: bold; color: rgb(153, 51, 153);">DNA</span>
sequences, using the HKY model:<br>
&nbsp;&nbsp; <span style="font-family: Courier New,Courier,monospace;">semphy
-s genes.fasta -o out.txt -T genes.tree -l log.txt -a&nbsp;4 --hky
--posterior</span><span
 style="font-family: Courier New,Courier,monospace;">DTME -O</span><br>
<span style="font-family: Courier New,Courier,monospace;"><span
 style="font-family: Times New Roman,Times,serif;">(Alphabet of 4
indicates DNA or RNA)</span><br>
</span></p>
<p>
</p>
<p>Same thing, with
100 <span
 style="font-weight: bold; font-style: italic; color: rgb(153, 51, 153);">bootstrap</span>
iterations:<br>
&nbsp;&nbsp; <span style="font-family: Courier New,Courier,monospace;">semphy
-s genes.fasta -o out.txt -T genes.tree -l log.txt -a&nbsp;4 --hky
--posterior</span><span
 style="font-family: Courier New,Courier,monospace;">DTME</span><span
 style="font-family: Courier New,Courier,monospace;"> -O -b 100</span></p>
<p><span style="text-decoration: underline;"><span
 style="font-weight: bold;">Maximum Likelihood (ML) trees using SEMPHY</span></span><span
 style="font-weight: bold; text-decoration: underline;"></span><span
 style="font-style: italic;"><br>
</span>Running SEMPHY to find the ML tree for a set of <span
 style="font-weight: bold; font-style: italic; color: rgb(153, 51, 153);">protein
sequences</span> using the JTT replacement matrix:&nbsp; (Standard NJ
will be used)<br>
&nbsp;&nbsp; <span style="font-family: Courier New,Courier,monospace;">semphy
-s prots.fasta -o out.txt -T prots.tree -l log.txt -a 20 --jtt -S -O</span><br>
(Meaning:&nbsp; Run SEMPHY on a fasta sequence file and write outputs,
tree and log files.&nbsp; Use alphabet of 20, i.e., amino acids.&nbsp;
Use the JTT matrix.&nbsp; Do SEMPHY steps, -O requests optimization of
rate parameters)<br>
</p>
<p>Same thing, but using the <span
 style="font-style: italic; font-weight: bold; color: rgb(153, 51, 153);">new
iterative NJ </span>method for the initial tree:<br>
&nbsp;&nbsp; <span style="font-family: Courier New,Courier,monospace;">semphy
-s prots.fasta -o out.txt -T prots.tree -l log.txt -a 20 --jtt -S
--posterior</span><span
 style="font-family: Courier New,Courier,monospace;">DTME -O</span></p>
Running SEMPHY to find the ML tree for a set of <span
 style="font-weight: bold; font-style: italic; color: rgb(153, 51, 153);">DNA
sequences</span> using the HKY model:<br>
&nbsp;&nbsp; <span style="font-family: Courier New,Courier,monospace;">semphy
-s genes.fasta -o out.txt -T genes.tree -l log.txt -a 4 --hky -S </span><span
 style="font-family: Courier New,Courier,monospace;">--posteriorDTME -O</span><br>
(Meaning:&nbsp; Run SEMPHY on a fasta sequence file and write outputs,
tree and log files.&nbsp; Use alphabet of 20, i.e., amino acids.&nbsp;
Use the HKY model.&nbsp; Do SEMPHY steps.&nbsp; Use iterative NJ)
<p>Same thing, with
100 <span
 style="font-weight: bold; font-style: italic; color: rgb(153, 51, 153);">bootstrap</span>
iterations:<br>
&nbsp;&nbsp; <span style="font-family: Courier New,Courier,monospace;">semphy
-s genes.fasta -o out.txt -T genes.tree -l log.txt -a&nbsp;4 --hky
-S --posterior</span><span
 style="font-family: Courier New,Courier,monospace;">DTME</span><span
 style="font-family: Courier New,Courier,monospace;"> -O -b 100</span></p>
<span style="text-decoration: underline;"><span
 style="font-weight: bold;"></span></span>
<p><span style="text-decoration: underline;"><span
 style="font-weight: bold;">List of options and parameters<br>
</span></span>The following table lists most of the available options
and parameters (the full list can be printed by typing
'semphy -h' at the command prompt)</p>
<table id="Table1" align="center" border="1">
  <tbody>
    <tr>
      <td bgcolor="#ffefce" valign="top"> Flag </td>
      <td bgcolor="#ffefce" valign="top"> Full name </td>
      <td bgcolor="#ffefce" valign="top"> Description </td>
      <td bgcolor="#ffefce" valign="top"> Default</td>
    </tr>
    <tr>
      <td valign="top"> -h </td>
      <td style="vertical-align: top;">--help<br>
      </td>
      <td valign="top">Print help and exit<br>
      </td>
      <td valign="top"><br>
      </td>
      <td valign="top"><br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><br>
      </td>
      <td style="vertical-align: top;">--full-help<br>
      </td>
      <td style="vertical-align: top;">Print help, including advanced
options, and exit<br>
      </td>
      <td style="vertical-align: top;"><br>
      </td>
    </tr>
    <tr>
      <td valign="top"> -s [MSA file] </td>
      <td style="vertical-align: top;">--sequence<br>
      </td>
      <td valign="top"> The input sequence file. The following formats
are supported: Mase, Molphy, Phylip, Clustal,
Fasta </td>
      <td valign="top"> Obligatory</td>
    </tr>
    <tr>
      <td valign="top"> -t [tree file] </td>
      <td style="vertical-align: top;">--tree<br>
      </td>
      <td valign="top"> An initial input tree file (in Newick format) </td>
      <td valign="top"> Optional</td>
    </tr>
    <tr>
      <td valign="top"> -o [output file] </td>
      <td style="vertical-align: top;">--outputfile<br>
      </td>
      <td valign="top">File for general outputs<br>
      </td>
      <td valign="top"> Optional</td>
    </tr>
    <tr>
      <td valign="top"> -T [output tree file] </td>
      <td style="vertical-align: top;">--treeoutputfile<br>
      </td>
      <td valign="top"> Output of the final tree </td>
      <td valign="top"> Optional</td>
    </tr>
    <tr>
      <td valign="top">-l [log file]</td>
      <td style="vertical-align: top;">--Logfile<br>
      </td>
      <td valign="top">Log file</td>
      <td valign="top">Optional</td>
    </tr>
    <tr>
      <td style="vertical-align: top;">-v [verbosity level]<br>
      </td>
      <td style="vertical-align: top;">--verbose<br>
      </td>
      <td style="vertical-align: top;">Verbosity level of the log file
(between 0 and 10)<br>
      </td>
      <td style="vertical-align: top;">Optional<br>
      </td>
    </tr>
    <tr>
      <td valign="top"> -a [alphabet] </td>
      <td style="vertical-align: top;">--alphabet<br>
      </td>
      <td valign="top">
      <p> 4 - nucleotides;&nbsp; 20 -&nbsp;amino acids;&nbsp; 61 -
codons </p>
      </td>
      <td valign="top"> 20</td>
    </tr>
    <tr>
      <td valign="top"><br>
      </td>
      <td style="vertical-align: top;">--BPrepeats<br>
      </td>
      <td valign="top">Perform a number of bootstrap iterations</td>
      <td valign="top">Optional</td>
    </tr>
    <tr>
      <td style="vertical-align: top;">-S<br>
      </td>
      <td style="vertical-align: top;">--SEMPHY<br>
      </td>
      <td style="vertical-align: top;">Do SEMPHY steps to search for
the ML tree<br>
      </td>
      <td style="vertical-align: top;">Optional<br>
      </td>
    </tr>
    <tr>
      <td style="vertical-align: top;"><strong>Distance Table
Estimation Method (DTME)</strong></td>
      <td style="vertical-align: top;"><br>
      </td>
      <td style="vertical-align: top;">Choice of NJ variant to be used
in SEMPHY, or by itself.&nbsp; Specifies the method that will be used
in NJ to calculate the distances table.&nbsp; Standard NJ is
-J.&nbsp; The recommended iterative NJ method is --posteriorDTME.<u><br>
      <span style="font-weight: bold;">Simple pairwise methods:</span></u><br>
-J is standard NJ, using ML distance with a
homogeneous rates model (also evoked by --homogeneousRatesDTME).<br>
ML distance with a
Gamma-ASRV model (--pairwiseGammaDTME) is usually not recommended.<br>
      <u style="font-weight: bold;">Iterative distance-based tree
reconstruction methods:</u><br>
Using the common alpha parameter (--commonAlphaDTME)<br>
Using the ML rate for each site (--rate4siteDTME)<br>
Using the posterior distribution of the rate at each site
(--posteriorDTME)</td>
      <td style="vertical-align: top;">NJ is not run unless some method
was requested, or if -S is used with no method chosen then -J is implied</td>
    </tr>
    <tr>
      <td valign="top"><strong>Evolutionary model</strong></td>
      <td style="vertical-align: top;"><br>
      </td>
      <td valign="top">
      <p>The following models are supported:<br>
JC for nucleotides (--nucjc) or amino acids (--aaJC), <br>
K2P (--k2p), HKY (--hky), Dayhoff (--day), JTT (--jtt), <br>
REV (--rev), WAG&nbsp;(--wag),&nbsp;cpREV (--cprev)<br>
Or load a matrix from a file:&nbsp; --modelfile=[file name]</p>
      </td>
      <td valign="top">JTT</td>
    </tr>
    <tr>
      <td valign="top"><strong><u>15acdfgilswAmong Site Rate Variation</u></strong></td>
      <td style="vertical-align: top;"><br>
      </td>
      <td valign="top"><br>
      </td>
      <td valign="top"><span
 style="font-weight: bold; text-decoration: underline;">NOTE:</span>&nbsp;
Either -H, -A, or -O must be given<br>
      </td>
    </tr>
    <tr>
      <td valign="top">-H</td>
      <td style="vertical-align: top;">--homogeneous<br>
      </td>
      <td valign="top">A homogeneous rates model (no Gamma ASRV)</td>
      <td valign="top">See above<br>
      </td>
    </tr>
    <tr>
      <td valign="top">-A [alpha]</td>
      <td style="vertical-align: top;">--alpha<br>
      </td>
      <td valign="top">Set the initial alpha parameter for Gamma ASRV</td>
      <td valign="top">See above</td>
    </tr>
    <tr>
      <td valign="top">-O</td>
      <td style="vertical-align: top;">--optimizeAlpha<br>
      </td>
      <td valign="top">Optimize the alpha parameter for the
reconstructed tree</td>
      <td valign="top">See above</td>
    </tr>
    <tr>
      <td valign="top"> -C [categories number] </td>
      <td style="vertical-align: top;">--categories<br>
      </td>
      <td valign="top"> The number of discrete categories used in the
approximation of the Gamma distribution of rates<br>
      </td>
      <td valign="top"> 8</td>
    </tr>
  </tbody>
</table>
<p>
</p>
<p> </p>
<span style="text-decoration: underline;"><span
 style="font-weight: bold;">Running on large datasets<br>
</span></span>Starting from version 2.0 SEMPHY can be run on very large
datasets of many thousands of sequences.&nbsp; However, in order to
handle such datasets a different make of the program must be used.
SEMPHY needs to be compiled with the doubleRep flag:<br>
<span style="font-family: Courier New,Courier,monospace;">&gt; make
doubleRep all</span><br>
This compilation command makes a copy of SEMPHY called <span
 style="font-family: Courier New,Courier,monospace;">semphy.doubleRep</span>.
This make of the program uses a&nbsp; different implementation of the
double data type that allows the handling of a virtually unlimited
number of sequences. The price for&nbsp; using doubleRep is a slower
running time by about one order of magnitude. Therefore, it is
recommended to use the normal make of SEMPHY where possible, and only
use the doubleRep make for datasets of 300 sequences or more.<br>
</body>
</html>
